src/meta/processors/zone/DropHostsProcessor.cpp

```diff
@@ -0,0 +1,156 @@
+    LOG(ERROR) << "List zones failed, error: " << apache::thrift::util::enumNameSafe(retCode);
+    handleErrorCode(retCode);
+    onFinished();
+    return;
+  }
+
+  auto iter = nebula::value(iterRet).get();
+  while (iter->valid()) {
+    auto hs = MetaKeyUtils::parseZoneHosts(iter->val());
+    if (std::all_of(hs.begin(), hs.end(), [&hosts](auto& h) {
+          return std::find(hosts.begin(), hosts.end(), h) != hosts.end();
+        })) {
+      auto zoneName = MetaKeyUtils::parseZoneName(iter->key());
+      LOG(INFO) << "Drop zone " << zoneName;
+      code = checkRelatedSpace(zoneName);
+      if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+        break;
+      }
+      data.emplace_back(iter->key());
+    }
+    iter->next();
+  }
+
+  // Detach the machine from cluster
+  for (auto& host : hosts) {
+    auto machineKey = MetaKeyUtils::machineKey(host.host, host.port);
+    auto ret = machineExist(machineKey);
+    if (ret != nebula::cpp2::ErrorCode::SUCCEEDED) {
+      LOG(ERROR) << "The host " << HostAddr(host.host, host.port) << " not existed!";
+      code = nebula::cpp2::ErrorCode::E_NO_HOSTS;
+      break;
+    }
+    data.emplace_back(std::move(machineKey));
+  }
+
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    handleErrorCode(code);
+    onFinished();
+    return;
+  }
+
+  resp_.set_code(nebula::cpp2::ErrorCode::SUCCEEDED);
+  doMultiRemove(std::move(data));
+}
+
+nebula::cpp2::ErrorCode DropHostsProcessor::checkRelatedSpace(const std::string& zoneName) {
+  const auto& prefix = MetaKeyUtils::spacePrefix();
+  auto ret = doPrefix(prefix);
+  if (!nebula::ok(ret)) {
+    auto retCode = nebula::error(ret);
+    LOG(ERROR) << "List spaces failed, error " << apache::thrift::util::enumNameSafe(retCode);
+    return nebula::cpp2::ErrorCode::E_KEY_NOT_FOUND;
+  }
+
+  auto iter = nebula::value(ret).get();
+  while (iter->valid()) {
+    auto properties = MetaKeyUtils::parseSpace(iter->val());
+    size_t replicaFactor = properties.get_replica_factor();
+    auto zones = properties.get_zone_names();
+    LOG(INFO) << "replica_factor " << replicaFactor << " zone size " << zones.size();
+    auto it = std::find(zones.begin(), zones.end(), zoneName);
+    if (it != zones.end()) {
+      if (zones.size() == replicaFactor) {
+        LOG(ERROR) << "Zone size is same with replica factor";
+        return nebula::cpp2::ErrorCode::E_CONFLICT;
+      } else {
+        zones.erase(it);
+        properties.set_zone_names(zones);
+        rewriteSpaceProperties(iter->key().data(), properties);
+      }
+    }
+    iter->next();
+  }
+  return nebula::cpp2::ErrorCode::SUCCEEDED;
+}
+
+void DropHostsProcessor::rewriteSpaceProperties(const std::string& spaceKey,
+                                                const meta::cpp2::SpaceDesc& properties) {
+  auto spaceVal = MetaKeyUtils::spaceVal(properties);
+  std::vector<kvstore::KV> data = {{spaceKey, spaceVal}};
+  doSyncPutAndUpdate(std::move(data));
```

Comment:
Don't write directly, you should make sure all writes succeed or failed.

Reply:
fine